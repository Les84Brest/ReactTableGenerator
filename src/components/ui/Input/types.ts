import { InputValidationCb } from '../Select/types';

export type Validation = {
    [idx: string]: number | boolean
}

export interface InputProps {
    id: string,
    className?: string,
    validations: Validation,
    validationStatus: InputValidationCb,
    value: string,
    onChange: (e: React.ChangeEvent<HTMLInputElement>) => void,
    resetInput: boolean,
    type: string,
    placeholder: string
}