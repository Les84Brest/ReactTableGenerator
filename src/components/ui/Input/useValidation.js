import { useState, useEffect, useRef } from "react";


export default function useValidation(value, validations) {

    // if validation rule value return true it means check falls

    const [isEmpty, setIsEmpty] = useState(false);
    const [minWidthError, setMinWidthError] = useState(false);
    const [lessThan, setLessThan] = useState(false);
    const [moreThan, setMoreThan] = useState(false);

    const errorMsgArr = useRef([])

    useEffect(() => {
        errorMsgArr.current = []
        for (const validation in validations) {

            switch (validation) {

                case 'isEmpty':
                    if (value) {
                        setIsEmpty(false)
                        break
                    }

                    errorMsgArr.current.push('Empty value')
                    setIsEmpty(true)

                    break
                case 'minWidth':
                    if (value.length < validations[validation]) {
                        errorMsgArr.current.push(`Length > ${validations[validation]}`)
                        setMinWidthError(true)

                        break
                    }
                    setMinWidthError(false)
                    break
                case 'minValue':
                    if (value && parseInt(value) < validations[validation]) {
                        setLessThan(true)
                        errorMsgArr.current.push(`Age > ${validations[validation]}`)
                        break
                    }
                    setLessThan(false)
                    break
                case 'maxValue':
                    if (value && parseInt(value) > validations[validation]) {
                        setMoreThan(true)
                        errorMsgArr.current.push(`Age < ${validations[validation]}`)
                        break
                    }
                    setMoreThan(false)
                    break
                default:
                    break;
            }
        }
    }, [value])

    const errorMessageText = errorMsgArr.current.join('. ')

    const isValid = !isEmpty && !minWidthError && !moreThan && !lessThan

    return { isValid, errorMessageText }
}