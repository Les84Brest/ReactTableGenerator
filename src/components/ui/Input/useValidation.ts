import { useState, useEffect, useRef } from "react";
import { Validation } from "./types";


export default function useValidation(
    value: string ,
    validations: Validation) {

    // if validation rule value return true it means check falls

    const [isEmpty, setIsEmpty] = useState<boolean>(false);
    const [minWidthError, setMinWidthError] = useState<boolean>(false);
    const [lessThan, setLessThan] = useState<boolean>(false);
    const [moreThan, setMoreThan] = useState<boolean>(false);

    const errorMsgArr = useRef<Array<string>>([])

    useEffect(() => {
        errorMsgArr.current = []
        for (const validation in validations) {

            switch (validation) {

                case 'isEmpty':
                    if (value) {
                        setIsEmpty(false)
                        break
                    }

                    errorMsgArr.current.push('Empty value')
                    setIsEmpty(true)

                    break
                case 'minWidth':
                    if ((value as string).length < parseInt(validations[validation])) {
                        errorMsgArr.current.push(`Length > ${validations[validation]}`)
                        setMinWidthError(true)

                        break
                    }
                    setMinWidthError(false)
                    break
                case 'minValue':
                    if (value && parseInt(value as string) < parseInt(validations[validation])) {
                        setLessThan(true)
                        errorMsgArr.current.push(`Age > ${validations[validation]}`)
                        break
                    }
                    setLessThan(false)
                    break
                case 'maxValue':
                    if (value && parseInt(value) > parseInt(validations[validation])) {
                        setMoreThan(true)
                        errorMsgArr.current.push(`Age < ${validations[validation]}`)
                        break
                    }
                    setMoreThan(false)
                    break
                default:
                    break;
            }
        }
    }, [value, validations])

    const errorMessageText = errorMsgArr.current.join('. ')

    const isValid = !isEmpty && !minWidthError && !moreThan && !lessThan

    return { isValid, errorMessageText }
}