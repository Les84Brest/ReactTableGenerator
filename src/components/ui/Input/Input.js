import React, { useState, useEffect } from "react"
import PropTypes from 'prop-types'
import classNames from 'classnames'
import useValidation from "./useValidation"

export const Input = ({
    id, className,resetInput, validations, onChange, value, validationStatus, ...attrs
}) => {

    const { isValid, errorMessageText } = useValidation(value, validations)
    const [isUsed, setIsUsed] = useState(false);

    const classes = classNames('form-input', className)

    useEffect(() => {
        validationStatus(isValid)
    }, [isValid])

    //reset Input state for new usage
    useEffect(()=> {
        resetInput && setIsUsed(false)
    }, [resetInput])

    const handleOnBlur = () => { setIsUsed(true) }

    return (

        <div className="form-input__wrap">
            <input
                name={id}
                className={classes}
                {...attrs}
                onBlur={handleOnBlur}
                value={value}
                onChange={onChange}
            />
            {(isUsed && !isValid) &&
                <span className="form-input__validation">{errorMessageText}</span>}
        </div>
    )
}
export default Input

Input.propTypes = {
    id: PropTypes.string.isRequired,
    className: PropTypes.string,
    validations: PropTypes.object,
    validationStatus: PropTypes.func,
    value: PropTypes.string,
    onChange: PropTypes.func,
    resetInput: PropTypes.bool
}

Input.defaultProps = {
    className: ''
}

