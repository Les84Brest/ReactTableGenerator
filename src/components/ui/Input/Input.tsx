import { useState, useEffect, FC } from "react"
import cn from 'classnames'
import useValidation from "./useValidation"
import { InputProps } from "./types"


export const Input: FC<InputProps> = ({
    id, className, resetInput, validations, onChange, value, validationStatus, ...attrs
}) => {

    const { isValid, errorMessageText } = useValidation(value, validations)
    const [isUsed, setIsUsed] = useState(false);

    const classes = cn('form-input', className)

    useEffect(() => {
        validationStatus(isValid)
    }, [isValid, validationStatus])

    //reset Input state for new usage
    useEffect(() => {
        resetInput && setIsUsed(false)
    }, [resetInput])

    const handleOnBlur = () => { setIsUsed(true) }

    return (
        <div className="form-input__wrap">
            <input
                name={id}
                className={classes}
                {...attrs}
                onBlur={handleOnBlur}
                value={value}
                onChange={onChange}
            />
            {(isUsed && !isValid) &&
                <span className="form-input__validation">{errorMessageText}</span>}
        </div>
    )
}
export default Input

