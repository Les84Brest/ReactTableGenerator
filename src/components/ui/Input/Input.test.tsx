import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Input from './Input';
import exp from 'constants';

const inputProps = {
    id: 'testId',
    className: 'test classes',
    validations:  {minWidth: 5},
    validationStatus: () =>{},
    value: 'test value',
    onChange: (e: React.ChangeEvent<HTMLInputElement>) => 'test changed',
    resetInput: false,
    type: 'text',
    placeholder: 'test placeholder'
}

describe('Input tests', () => {

    test('conponent renders', () => {

        render(<Input {...inputProps} />)
        const input = screen.getByPlaceholderText(/test placeholder/i)

        expect(input).toBeInTheDocument();
        expect(input).toMatchSnapshot();
    })

    // test('type text', () => {

    //     render(<Input {...inputProps} />)
    //     const input = screen.getByPlaceholderText(/test placeholder/i)

    //     userEvent.type(input, 'testText');

    //     expect(screen.getByPlaceholderText(/test placeholder/i)).toContainHTML('testText');
    // })

});