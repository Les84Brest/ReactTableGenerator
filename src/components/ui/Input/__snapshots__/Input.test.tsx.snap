// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input tests conponent renders 1`] = `
<input
  class="form-input test classes"
  name="testId"
  placeholder="test placeholder"
  type="text"
  value="test value"
/>
`;
